# 1. Phylogenetic tree
megacc -a ./megacc_nj.mao -d all_0.05_0.1.fa -f Fasta

```
i; Please do not edit this file! If this file is modified, results are unpredictable.; Instead of modifying this file, simply create a new MEGA Analysis Options file by using MEGA.
[ MEGAinfo ]
ver                                = 10200331-x86_64             
[ DataSettings ]
datatype                           = snNucleotide                      
containsCodingNuc                  = False                             
MissingBaseSymbol                  = N                                 
IdenticalBaseSymbol                = .                                 
GapSymbol                          = -                                 
Labelled Sites                     = All Sites                         
Labels to Include                  =                                   
[ ProcessTypes ]
ppInfer                            = true                              
ppNJ                               = true                              
[ AnalysisSettings ]
Analysis                           = Phylogeny Reconstruction          
Scope                              = All Selected Taxa                 
Statistical Method                 = Neighbor-joining                  
Phylogeny Test                     = ====================              
Test of Phylogeny                  = Bootstrap method                  
No. of Bootstrap Replications      = 100                              
Substitution Model                 = ====================              
Substitutions Type                 = Nucleotide                        
Model/Method                       = Maximum Composite Likelihood      
Substitutions to Include           = d: Transitions + Transversions    
Rates and Patterns                 = ====================              
Rates among Sites                  = Uniform Rates                     
Gamma Parameter                    = Not Applicable                    
Pattern among Lineages             = Same (Homogeneous)                
Data Subset to Use                 = ====================              
Gaps/Missing Data Treatment        = Pairwise deletion                 
Site Coverage Cutoff (%)           = Not Applicable                    
System Resource Usage              = ====================              
Number of Threads                  = 60                                 
Has Time Limit                     = False                             
Maximum Execution Time             = -1 
```

# 2. Population structure
for i in {2..20}; do admixture -s time --cv /datalus/heqiang/millet/analysis/evolution/data/NGS/1844/all_0.05_0.1.bed ${i} | tee log${i}.out; done
